cmd/encryption-v1.go

```diff
@@ -161,51 +198,76 @@ func DecryptRequest(client io.Writer, r *http.Request, metadata map[string]strin
 	}
 	delete(metadata, SSECustomerKey) // make sure we do not save the key by accident
 
-	if metadata[ContentEncryptionAlgorithm] != DareHmacSha256 {
-		return nil, errInvalidSSEAlgorithm // TODO(aead): different error
+	if metadata[ServerSideEncryptionKDF] != SSEKeyDerivationHmacSha256 { // currently HMAC-SHA256 is the only option
+		return nil, errObjectTampered
 	}
-	nonce, err := base64.StdEncoding.DecodeString(metadata[ClientSideEncryptionIV])
-	if err != nil {
+	nonce, err := base64.StdEncoding.DecodeString(metadata[ServerSideEncryptionIV])
+	if err != nil || len(nonce) != 32 {
+		return nil, errObjectTampered
+	}
+	keyMAC, err := base64.StdEncoding.DecodeString(metadata[ServerSideEncryptionKeyMAC])
+	if err != nil || len(keyMAC) != 32 {
 		return nil, errObjectTampered
 	}
 
-	mac := hmac.New(sha256.New, key)
+	keyHash := sha256.Sum256(key) // verify that client provided correct key
+	mac := hmac.New(sha256.New, keyHash[:])
 	mac.Write(nonce)
+	if !hmac.Equal(keyMAC, mac.Sum(nil)) {
+		return nil, errSSEKeyMismatch // client-provided key is wrong or object metadata was modified
+	}
+
+	mac = hmac.New(sha256.New, key) // derive decryption key
+	mac.Write(keyMAC)
 	writer, err := sio.DecryptWriter(client, sio.Config{Key: mac.Sum(nil)})
 	if err != nil {
 		return nil, errInvalidSSEKey
 	}
-	delete(metadata, ClientSideEncryptionIV)
-	delete(metadata, ContentEncryptionAlgorithm)
+
+	delete(metadata, ServerSideEncryptionIV)
+	delete(metadata, ServerSideEncryptionKDF)
+	delete(metadata, ServerSideEncryptionKeyMAC)
 	return writer, nil
 }
 
-// DecryptSSECustomerObjectInfo converts an object info of an SSE-C encrypted object into
-// a plain object info. It returns an error if the object info is not a validly encrypted
-// object info.
-//
-// DecryptSSECustomerObjectInfo computes the real object size of an encrypted object.
-func DecryptSSECustomerObjectInfo(info ObjectInfo) (ObjectInfo, error) {
-	size := (info.Size / (32 + 64*1024)) * (64 * 1024)
-	if mod := info.Size % (32 + 64*1024); mod > 0 {
-		if mod < 33 {
-			return info, errObjectTampered
+// IsEncrypted returns true if the object is marked as encrypted.
+func (o *ObjectInfo) IsEncrypted() bool {
+	if _, ok := o.UserDefined[ServerSideEncryptionIV]; ok {
+		return true
+	}
+	if _, ok := o.UserDefined[ServerSideEncryptionKDF]; ok {
+		return true
+	}
+	if _, ok := o.UserDefined[ServerSideEncryptionKeyMAC]; ok {
+		return true
+	}
+	return false
+}
+
+// DecryptedSize returns the size of the object after decryption in bytes.
+// It returns an error if the object is not encrypted or marked as encrypted
+// but has an invalid size.
+func (o *ObjectInfo) DecryptedSize() (int64, error) {
+	if !o.IsEncrypted() {
+		return -1, errObjectTampered // TODO(aead): think about that...
```

Comment:
what we do think about? this shouldn't be called based on the contract so caller if they are calling they are on their own and the decryption will fail, we don't need to return an error here IMHO.

Reply:
Also objectIsNotTampered is incorrect - we can say objectNotEcnrypted.

Reply:
Exactly that were my thoughts ^^
Good catch - Okay so your saying we can ignore the case that the object may not be encrypted?
From my point of view a `panic` would be appropriate but I think we agreed on: no panic in codebase, right?

Reply:
A panic is fine in this i think, since its unexpected usage of the API itself and potentially would warn developers directly. @aead

Reply:
Agree - okay, will add panic 👍